compiler/src/codegen/compcore.re

```diff
@@ -3387,66 +3200,95 @@ let compile_globals = (wasm_mod, env, {globals} as prog) => {
       Global.add_global(
         wasm_mod,
-        Ident.unique_name(id),
-        wasm_type(ty),
-        true,
-        Expression.Const.make(wasm_mod, initial_value(ty)),
-      ),
+        name,
+        wasm_type(allocation_type),
+        mutable_,
+        switch (initial) {
+        | Some(initial) =>
+          Expression.Const.make(wasm_mod, compile_const(initial))
+        | None =>
+          Expression.Const.make(wasm_mod, initial_value(allocation_type))
+        },
+      );
+    },
     globals,
   );
-  ignore @@
-  Global.add_global(
-    wasm_mod,
-    Ident.name(table_size),
-    Type.int32,
-    false,
-    Expression.Const.make(
-      wasm_mod,
-      const_int32(compute_table_size(env, prog)),
-    ),
-  );
 };
 
 let compile_main = (wasm_mod, env, prog) => {
-  ignore @@
-  compile_function(
-    ~name=grain_main,
-    wasm_mod,
-    env,
-    {
-      id: Ident.create(grain_main),
-      name: Some(grain_main),
-      args: [],
-      return_type: [Types.Unmanaged(WasmI32)],
-      closure: None,
-      body: prog.main_body,
-      stack_size: prog.main_body_stack_size,
-      attrs: [],
-      func_loc: Grain_parsing.Location.dummy_loc,
+  let num_mains = List.length(prog.programs);
+  List.iteri(
+    (dep_id, prog: mash_program) => {
+      let env = {...env, compilation_mode: prog.compilation_mode, dep_id};
+      let compile = () => {
+        ignore @@
+        compile_function(
+          ~name=grain_main,
+          wasm_mod,
+          env,
+          {
+            id: Ident.create(grain_main),
+            name: Some(grain_main),
+            args: [],
+            return_type: [Types.Unmanaged(WasmI32)],
+            closure: None,
+            body: prog.main_body,
+            stack_size: prog.main_body_stack_size,
+            attrs: [],
+            func_loc: Grain_parsing.Location.dummy_loc,
+          },
+        );
+      };
+      switch (prog.compilation_mode) {
+      | Runtime =>
+        Config.preserve_config(() => {
+          Config.no_gc := true;
+          compile();
+        })
+      | Normal => compile()
+      };
     },
+    prog.programs,
   );
-  if (! Config.use_start_section^) {
-    ignore @@
+  let starts =
+    List.init(num_mains, dep_id =>
+      Expression.Drop.make(
+        wasm_mod,
+        Expression.Call.make(
+          wasm_mod,
+          Printf.sprintf("%s_%d", grain_main, dep_id),
+          [],
+          Type.int32,
+        ),
+      )
+    );
+  let start =
     Function.add_function(
       wasm_mod,
       grain_start,
       Type.none,
       Type.none,
       [||],
-      Expression.Drop.make(
-        wasm_mod,
-        Expression.Call.make(wasm_mod, grain_main, [], Type.int32),
-      ),
+      Expression.Block.make(wasm_mod, "_start", starts),
     );
+  if (Grain_utils.Config.use_start_section^) {
+    Function.set_start(wasm_mod, start);
+  } else {
+    ignore @@
+    Export.add_function_export(wasm_mod, "_start", Comp_utils.grain_start);
```

Comment:
Nit: inconsistent use of `grain_start` constant and "_start" literal (a few other times in the lines above)